assets/tests/consolidated_1_1_tests.yaml

```diff
@@ -5090,4 +5090,30 @@ tests:
               user: user:jon
               type: document
               relation: viewer
-            errorCode: 2002
\ No newline at end of file
+            errorCode: 2002
+  - name: userset_defines_itself
+    stages:
+      - model: |
+          type user
+
+          type document
+            relations
+              define viewer: [user] as self
+        tuples:
+        checkAssertions:
+          - tuple:
+              user: document:1#viewer
+              relation: viewer
+              object: document:1
```

Comment:
should we be rejecting these tuples on writes since they are redundant?

Reply:
This tuple wouldn't be a valid tuple at Write time, because the type restrictions doesn't allow writing `document#viewer` relationships to `document#viewer`. This is currently how the system behaves, and that should remain unchanged. This behavior is a reflexive property of a set of users in general. A set of objects related to the object itself is self defining. That is

`document:1#viewer` reflects itself `document:1#viewer`. So if you Check it, then it is logical that it should be {allowed: true}.

Reply:
rephrasing my question: if the model did allow such tuples, e.g.

model
  schema 1.1
type doc
	 viewer: [doc#viewer]


then should the write of `doc:1 # viewer @ doc:1#viewer`  be allowed? we're going to consider this tuple as `true` and never look it up 🤷‍♀️

Reply:
Yes, there's nothing restricting that from occurring and nor should there be. If you want to establish a concrete relationship with a userset of the same type and relation, then you should be able to do so.  For example, if you want to establish that all subjects who can view `doc:Z` can view `doc:1` (e.g. `doc:1#viewer@doc:Z#viewer`) then you should be able to do that. By extension, if you want to establish a self-defining tuple of the form `doc:1#viewer@doc:1#viewer` then writing that tuple is reasonable so long as the type restrictions allow it as well.

The invariants for Check are different than the invariants for Write. Namely, you can Check if a subject has a relationship who would otherwise be unrelated based on the type-restrictions, and you'll get an `{allowed: false}` outcome. Whereas with Write you can only write that which the type restrictions permit.

If you want to determine if the set of subjects of some relationship contain the same set of subjects from another relationship, then that is also reasonable to ask OpenFGA (e.g. `Check(document:1#viewer@group:eng#member)`) . By extension, if you want to Check if the set of subjects of some relationship are also related to that same relationship, then the answer should always be `{allowed: true}` because of the reflexive property of sets. A set of subjects of some relationships is equal to the same set of subjects of the same relationship. For example, `Check(document:1#viewer@document:1#viewer) --> {allowed: true}`.